File: tensorflow/core/kernels/composite_tensor_ops.cc
/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "tensorflow/core/framework/op.h"
#include "tensorflow/core/framework/op_kernel.h"
#include "tensorflow/core/framework/variant.h"
#include "tensorflow/core/framework/variant_encode_decode.h"
#include "tensorflow/core/kernels/composite_tensor_variant.h"
#include "tensorflow/core/platform/errors.h"
#include "tensorflow/core/protobuf/composite_tensor_variant.pb.h"
#include "tensorflow/core/protobuf/struct.pb.h"

namespace tensorflow {

class CompositeTensorVariantFromComponents : public OpKernel {
 public:
  explicit CompositeTensorVariantFromComponents(OpKernelConstruction* context)
      : OpKernel(context) {
    string type_spec_string;
    OP_REQUIRES_OK(context, context->GetAttr("metadata", &type_spec_string));
    OP_REQUIRES(context, metadata_.ParseFromString(type_spec_string),
                errors::InvalidArgument("Error parsing metadata"));
  }

  void Compute(OpKernelContext* context) override {
    OpInputList components_in;
    OP_REQUIRES_OK(context, context->input_list("components", &components_in));

    Tensor* encoded;
    OP_REQUIRES_OK(context,
                   context->allocate_output(0, TensorShape({}), &encoded));

    std::vector<Tensor> components{components_in.begin(), components_in.end()};
    encoded->flat<Variant>()(0) =
        CompositeTensorVariant(metadata_, absl::MakeSpan(components));
  }

 private:
  CompositeTensorVariantMetadata metadata_;
};

class CompositeTensorVariantToComponents : public OpKernel {
 public:
  explicit CompositeTensorVariantToComponents(OpKernelConstruction* context)
      : OpKernel(context) {
    string type_spec_string;
    OP_REQUIRES_OK(context, context->GetAttr("metadata", &type_spec_string));
    OP_REQUIRES(context, metadata_.ParseFromString(type_spec_string),
                errors::InvalidArgument("Error parsing `metadata`"));

    OP_REQUIRES_OK(context,
                   context->GetAttr("Tcomponents", &component_dtypes_));
  }

  void Compute(OpKernelContext* context) override {
    Tensor encoded_t = context->input(0);
    auto* encoded = encoded_t.flat<Variant>()(0).get<CompositeTensorVariant>();

    // Check that the encoded TypeSpec is compatible with the expected TypeSpec.
    // For now, we just check that the class matches.
    //
    // TODO(b/173744905): Update this to do a generic compatibility check. This
    // would require replacing the current design, where Python subclasses of
    // TypeSpec can override is_compatible, with a design where compatibility
    // can be deterministically determined from the metadata.
    auto expected_class = metadata_.type_spec_proto().type_spec_class();
    auto actual_class = encoded->metadata().type_spec_proto().type_spec_class();
    OP_REQUIRES(
        context, expected_class == actual_class,
        errors::InvalidArgument(
            "Expected a ", TypeSpecProto::TypeSpecClass_Name(expected_class),
            " (based on `type_spec`), but `encoded` contains a ",
            TypeSpecProto::TypeSpecClass_Name(actual_class)));

    // Extract the component tensors.
    OpOutputList components;
    OP_REQUIRES_OK(context, context->output_list("components", &components));
    int num_components = encoded->flat_components().size();

    OP_REQUIRES(context, component_dtypes_.size() == num_components,
                errors::InvalidArgument("Encoded value has ", num_components,
                                        " tensor components; expected ",
                                        component_dtypes_.size(),
                                        " components based on type_spec"));

    for (int i = 0; i < component_dtypes_.size(); i++) {
      const Tensor& component = encoded->flat_components()[i];
      OP_REQUIRES(context, component_dtypes_[i] == component.dtype(),
                  errors::InvalidArgument("Tensor component ", i, " had dtype ",
                                          DataType_Name(component.dtype()),
                                          "; expected dtype ",
                                          DataType_Name(component_dtypes_[i])));
      components.set(i, component);
    }
  }

 private:
  CompositeTensorVariantMetadata metadata_;
  std::vector<DataType> component_dtypes_;
};

REGISTER_KERNEL_BUILDER(
    Name("CompositeTensorVariantToComponents").Device(DEVICE_CPU),
    CompositeTensorVariantToComponents);
REGISTER_KERNEL_BUILDER(
    Name("CompositeTensorVariantFromComponents").Device(DEVICE_CPU),
    CompositeTensorVariantFromComponents);

}  // namespace tensorflow

File: tensorflow/python/ops/composite_tensor_ops.py
# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# =============================================================================
"""Operations for ExtensionTypes (aka Composite Tensors)."""

from tensorflow.core.protobuf import composite_tensor_variant_pb2
from tensorflow.python.framework import composite_tensor
from tensorflow.python.framework import dtypes
from tensorflow.python.framework import ops
from tensorflow.python.ops import gen_composite_tensor_ops
from tensorflow.python.saved_model import nested_structure_coder
from tensorflow.python.util import nest


def composite_tensor_to_variants(value, type_spec=None, name=None):
  """Encodes `value` as a scalar variant tensor.

  Args:
    value: The `ExtensionType` value to encode.
    type_spec: Information about the value's type that should be included in the
      encoding.
    name: Optional name for the operation.

  Returns:
    A Tensor with shape=`()` and dtype=`tf.variant`.

  Raises:
    ValueError: If `type_spec` is not compatible with `value`.
  """
  if not isinstance(value, composite_tensor.CompositeTensor):
    raise TypeError("Expected `value` to be a CompositeTensor. "
                    f"Received {type(value)}.")

  if type_spec is None:
    type_spec = value._type_spec  # pylint: disable=protected-access
  if not type_spec.is_compatible_with(value):
    raise ValueError(f"`type_spec` {type_spec} is not compatible with `value` "
                     f"{value!r}.")
  metadata = composite_tensor_variant_pb2.CompositeTensorVariantMetadata()
  metadata.type_spec_proto.CopyFrom(
      nested_structure_coder.encode_structure(type_spec).type_spec_value)

  return gen_composite_tensor_ops.CompositeTensorVariantFromComponents(
      components=nest.flatten(value, expand_composites=True),
      metadata=metadata.SerializeToString(),
      name=name)


def composite_tensor_from_variant(encoded, type_spec, name=None):
  """Returns the `ExtensionType` value encoded by a variant scalar tensor.

  Args:
    encoded: A Tensor returned by `composite_tensor_to_variants`.
    type_spec: The `TypeSpec` of the original value.  This is used to determine
      the number and types of the component tensors that comprise the decoded
      value.  Must be compatible with the `TypeSpec` serilized in `encoded`.
    name: Optional name for the operation.

  Returns:
    An `ExtensionType` value that is compatible with `TypeSpec`.

  Raises:
    TypeError: If `encoded` is not a Tensor with dtype=variant.
    InvalidArgumentError: If `encoded` is not compatible with `type_spec`.
  """
  if not isinstance(encoded, ops.Tensor):
    raise TypeError(f"Expected `encoded` to be a Tensor, got {encoded!r}.")
  if encoded.dtype != dtypes.variant:
    raise TypeError("Expected `encoded` to have dtype=variant, got "
                    f"{encoded!r}.")
  encoded.shape.assert_is_compatible_with(())

  metadata = composite_tensor_variant_pb2.CompositeTensorVariantMetadata()
  metadata.type_spec_proto.CopyFrom(
      nested_structure_coder.encode_structure(type_spec).type_spec_value)

  component_dtypes = [
      t.dtype for t in nest.flatten(type_spec, expand_composites=True)
  ]

  components = gen_composite_tensor_ops.CompositeTensorVariantToComponents(
      encoded=encoded,
      metadata=metadata.SerializeToString(),
      Tcomponents=component_dtypes,
      name=name)
  return nest.pack_sequence_as(type_spec, components, expand_composites=True)


@ops.RegisterGradient("CompositeTensorVariantFromComponents")
def _composite_tensor_to_variants_grad(op, grad):
  return gen_composite_tensor_ops.CompositeTensorVariantToComponents(
      encoded=grad,
      metadata=op.get_attr("metadata"),
      Tcomponents=op.get_attr("Tcomponents"))


@ops.RegisterGradient("CompositeTensorVariantToComponents")
def _composite_tensor_from_variant_grad(op, *grad):
  assert len(grad) == len(op.outputs)
  # `components` is `op.outputs`, but with any tensors for which we're
  # taking the gradient replaced by the corresponding value from `grad`.
  components = [
      op.outputs[i] if grad[i] is None else grad[i] for i in range(len(grad))
  ]
  return gen_composite_tensor_ops.CompositeTensorVariantFromComponents(
      components=components, metadata=op.get_attr("metadata"))
